README.md
# Introducing Code Llama

Code Llama is a family of large language models for code based on [Llama 2](https://github.com/facebookresearch/llama) providing state-of-the-art performance among open models, infilling capabilities, support for large input contexts, and zero-shot instruction following ability for programming tasks. We provide multiple flavors to cover a wide range of applications: foundation models (Code Llama), Python specializations (Code Llama - Python), and instruction-following models (Code Llama - Instruct) with 7B, 13B and 34B parameters each. All models are trained on sequences of 16k tokens and show improvements on inputs with up to 100k tokens. 7B and 13B Code Llama and Code Llama - Instruct variants support infilling based on surrounding content. Code Llama was developed by fine-tuning Llama 2 using a higher sampling of code. As with Llama 2, we applied considerable safety mitigations to the fine-tuned versions of the model. For detailed information on model training, architecture and parameters, evaluations, responsible AI and safety refer to  our [research paper](https://ai.meta.com/research/publications/code-llama-open-foundation-models-for-code/). Output generated by code generation features of the Llama Materials, including Code Llama, may be subject to third party licenses, including, without limitation, open source licenses.

We are unlocking the power of large language models and our latest version of Code Llama is now accessible to individuals, creators, researchers and businesses of all sizes so that they can experiment, innovate and scale their ideas responsibly. This release includes model weights and starting code for pretrained and fine-tuned Llama language models — ranging from 7B to 34B parameters.

This repository is intended as a minimal example to load [Code Llama](https://ai.meta.com/research/publications/code-llama-open-foundation-models-for-code/) models and run inference.


[comment]: <> (Code Llama models are compatible with the scripts in llama-recipes)


## Download

In order to download the model weights and tokenizers, please visit the [Meta website](https://ai.meta.com/resources/models-and-libraries/llama-downloads/) and accept our License.

Once your request is approved, you will receive a signed URL over email. Then run the download.sh script, passing the URL provided when prompted to start the download. Make sure that you copy the URL text itself, **do not use the 'Copy link address' option** when you right click the URL. If the copied URL text starts with: https://download.llamameta.net, you copied it correctly. If the copied URL text starts with: https://l.facebook.com, you copied it the wrong way.

Pre-requisites: make sure you have `wget` and `md5sum` installed. Then to run the script: `bash download.sh`.

Keep in mind that the links expire after 24 hours and a certain amount of downloads. If you start seeing errors such as `403: Forbidden`, you can always re-request a link.

### Model sizes

|  Model | Size |
|--------|----|
| 7B     | ~12.55GB  |
| 13B    | 24GB  |
| 34B    | 63GB  |

[comment]: <> (Access on Hugging Face, We are also providing downloads on Hugging Face. You must first request a download from the Meta website using the same email address as your Hugging Face account. After doing so, you can request access to any of the models on Hugging Face and within 1-2 days your account will be granted access to all versions.)

## Setup

In a conda env with PyTorch / CUDA available, clone the repo and run in the top-level directory:

```
pip install -e .
```

## Inference

Different models require different model-parallel (MP) values:

|  Model | MP |
|--------|----|
| 7B     | 1  |
| 13B    | 2  |
| 34B    | 4  |

All models support sequence lengths up to 100,000 tokens, but we pre-allocate the cache according to `max_seq_len` and `max_batch_size` values. So set those according to your hardware and use-case.

### Pretrained Code Models

The Code Llama and Code Llama - Python models are not fine-tuned to follow instructions. They should be prompted so that the expected answer is the natural continuation of the prompt.


See `example_completion.py` for some examples. To illustrate, see command below to run it with the `CodeLlama-7b` model (`nproc_per_node` needs to be set to the `MP` value):

```
torchrun --nproc_per_node 1 example_completion.py \
    --ckpt_dir CodeLlama-7b/ \
    --tokenizer_path CodeLlama-7b/tokenizer.model \
    --max_seq_len 128 --max_batch_size 4
```

Pretrained code models are: the Code Llama models `CodeLlama-7b`, `CodeLlama-13b`, `CodeLlama-34b` and the Code Llama - Python models 
`CodeLlama-7b-Python`, `CodeLlama-13b-Python`, `CodeLlama-34b-Python`.

### Code Infilling

Code Llama and Code Llama - Instruct 7B and 13B models are capable of filling in code given the surrounding context.


See `example_infilling.py` for some examples. The `CodeLlama-7b` model can be run for infilling with the command below (`nproc_per_node` needs to be set to the `MP` value):
```
torchrun --nproc_per_node 1 example_infilling.py \
    --ckpt_dir CodeLlama-7b/ \
    --tokenizer_path CodeLlama-7b/tokenizer.model \
    --max_seq_len 192 --max_batch_size 4
```

Pretrained infilling models are: the Code Llama models `CodeLlama-7b` and `CodeLlama-13b` and the Code Llama - Instruct models `CodeLlama-7b-Instruct`, `CodeLlama-13b-Instruct`.

### Fine-tuned Instruction Models

Code Llama - Instruct models are fine-tuned to follow instructions. To get the expected features and performance for them, a specific formatting defined in [`chat_completion`](https://github.com/facebookresearch/codellama/blob/main/llama/generation.py#L279-L366)
needs to be followed, including the `INST` and `<<SYS>>` tags, `BOS` and `EOS` tokens, and the whitespaces and linebreaks in between (we recommend calling `strip()` on inputs to avoid double-spaces).
You can use `chat_completion` directly to generate answers with the instruct model. 

You can also deploy additional classifiers for filtering out inputs and outputs that are deemed unsafe. See the llama-recipes repo for [an example](https://github.com/facebookresearch/llama-recipes/blob/main/src/llama_recipes/inference/safety_utils.py) of how to add a safety checker to the inputs and outputs of your inference code.

Examples using `CodeLlama-7b-Instruct`:

```
torchrun --nproc_per_node 1 example_instructions.py \
    --ckpt_dir CodeLlama-7b-Instruct/ \
    --tokenizer_path CodeLlama-7b-Instruct/tokenizer.model \
    --max_seq_len 512 --max_batch_size 4
```

Fine-tuned instruction-following models are: the Code Llama - Instruct models `CodeLlama-7b-Instruct`, `CodeLlama-13b-Instruct`, `CodeLlama-34b-Instruct`.

Code Llama is a new technology that carries potential risks with use. Testing conducted to date has not — and could not — cover all scenarios.
In order to help developers address these risks, we have created the [Responsible Use Guide](https://github.com/facebookresearch/llama/blob/main/Responsible-Use-Guide.pdf). More details can be found in our research papers as well.

## Issues
Please report any software “bug”, or other problems with the models through one of the following means:
- Reporting issues with the model: [github.com/facebookresearch/codellama](http://github.com/facebookresearch/codellama)
- Reporting risky content generated by the model: [developers.facebook.com/llama_output_feedback](http://developers.facebook.com/llama_output_feedback)
- Reporting bugs and security concerns: [facebook.com/whitehat/info](http://facebook.com/whitehat/info)

## Model Card
See [MODEL_CARD.md](MODEL_CARD.md) for the model card of Code Llama.

## License

Our model and weights are licensed for both researchers and commercial entities, upholding the principles of openness. Our mission is to empower individuals, and industry through this opportunity, while fostering an environment of discovery and ethical AI advancements.

See the [LICENSE](https://github.com/facebookresearch/llama/blob/main/LICENSE) file, as well as our accompanying [Acceptable Use Policy](https://github.com/facebookresearch/llama/blob/main/USE_POLICY.md)

## References

1. [Code Llama Research Paper](https://ai.meta.com/research/publications/code-llama-open-foundation-models-for-code/)
2. [Code Llama Blog Post](https://ai.meta.com/blog/code-llama-large-language-model-coding/)
# codesummarizer README

This is the README for your extension "codesummarizer". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Working with Markdown

You can author your README using Visual Studio Code.  Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux)
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux)
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "codesummarizer" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

test_nodejs_communication.py
import subprocess
import os
import time
import platform
import win32pipe  # Import the win32pipe module from pywin32
import win32file

class TestNodeJSCommunication:
    def setUp(self):
        # Create pipes based on the operating system
        if platform.system() == 'Windows':
            self.create_named_pipe("python_to_js_pipe")
            # Note: 'js_to_python_pipe' is created by the Node.js script
        else:
            os.mkfifo("python_to_js_pipe")
            # Note: 'js_to_python_pipe' is created by the Node.js script

    def tearDown(self):
        # Remove pipes after the test
        os.remove("python_to_js_pipe")
        # Note: 'js_to_python_pipe' is removed by the Node.js script after the test

    def create_named_pipe(self, pipe_name):
        # Create named pipes on Windows using pywin32
        if platform.system() == 'Windows':
            win32pipe.CreateNamedPipe(fr'\\.\pipe\{pipe_name}', win32pipe.PIPE_ACCESS_DUPLEX,
                                      win32pipe.PIPE_TYPE_MESSAGE | win32pipe.PIPE_READMODE_MESSAGE | win32pipe.PIPE_WAIT,
                                      1, 65536, 65536, 300, None)
        else:
            os.mkfifo(pipe_name)

    def test_nodejs_communication(self):
        # Start the Node.js script as a subprocess
        js_process = subprocess.Popen(['node', 'node_script.js'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)

        try:
            # Wait for the 'js_to_python_pipe' to be created by the Node.js script
            time.sleep(2)  # Adjust the sleep duration based on your needs

            # Open pipes for reading and writing
            with open("python_to_js_pipe", 'wb') as python_to_js_pipe, open("js_to_python_pipe", 'rb') as js_to_python_pipe:
                # User input: Get a string from the user
                message_to_js = input("Enter a message to send to Node.js: ")
                python_to_js_pipe.write(message_to_js.encode('utf-8'))
                python_to_js_pipe.flush()

                # Wait for a response from Node.js
                time.sleep(5)  # You might want to use more robust synchronization methods

                # Read the response from Node.js
                message_from_js = js_to_python_pipe.read().decode('utf-8').strip()

                # Print the message received from Node.js
                print('Received message from Node.js:', message_from_js)

                # Assert that the processed message matches the expected response
                assert message_from_js == f"Processed: {message_to_js}"


        finally:
            # Terminate the Node.js subprocess
            js_process.terminate()

if __name__ == '__main__':
    test_instance = TestNodeJSCommunication()
    test_instance.setUp()
    test_instance.test_nodejs_communication()
    test_instance.tearDown()





extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require('vscode');
const { spawn } = require('child_process');

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// // This line of code will only be executed once when your extension is activated
	// console.log('Congratulations, your extension "codesummarizer" is now active!');

	// // The command has been defined in the package.json file
	// // Now provide the implementation of the command with  registerCommand
	// // The commandId parameter must match the command field in package.json
	// let disposable = vscode.commands.registerCommand('codesummarizer.helloWorld', function () {
	// 	// The code you place here will be executed every time your command is executed

	// 	// Display a message box to the user
	// 	vscode.window.showInformationMessage('Hello World from CodeSummarizer!');
	// });

	// context.subscriptions.push(disposable);
	
	let summarizer = spawn('python', ['path/to/your/script.py']);
	let disposable = vscode.window.onDidChangeTextEditorSelection(event => {
        // Check if the selection is non-empty (highlighted text)
        if (!event.selections[0].isEmpty) {
            const document = event.textEditor.document;
            const selection = event.selections[0];

            // Handle the highlighted text
            handleHighlight(document, selection);
        }
    });

    context.subscriptions.push(disposable);
}

function handleHighlight(document, selection) {
    // Logic to handle the highlighted text
}


	// let hoverProvider = vscode.languages.registerHoverProvider('*', {
    //     provideHover(document, position, token) {
    //         // Asynchronous operation wrapped in a promise
    //         const asyncOperation = new Promise((resolve, reject) => {
    //             fetchSummary(document, position, token, resolve, reject);
    //         });

    //         // Timeout promise
    //         const timeout = new Promise((resolve) => {
    //             const checkInterval = setInterval(() => {
    //                 if (token.isCancellationRequested) {
    //                     clearInterval(checkInterval);
    //                     resolve(null); // Resolve with null if cancelled
    //                 }
    //             }, 1000); // Check every 1 second for cancellation

    //             // Set a maximum time for the operation (e.g., 10 seconds)
    //             setTimeout(() => {
    //                 clearInterval(checkInterval);
    //                 resolve(null); // Resolve with null on timeout
    //             }, 10000);
    //         });

    //         // Race between async operation and timeout
    //         return Promise.race([asyncOperation, timeout]);
    //     }
    // });


    // context.subscriptions.push(hoverProvider);
// }

function fetchSummary(document, position, token, resolve, reject) {
    // Perform the actual async task
    // Example: Fetch data, perform calculations, etc.
    // Resolve the promise when the operation is complete
    // Example: resolve(new vscode.Hover("Hover Information"));
}

// This method is called when your extension is deactivated
function deactivate() {}

module.exports = {
	activate,
	deactivate
}


//git remote add project-a /path/to/project-a
// git fetch project-a --tags
// git merge --allow-unrelated-histories project-a/master # or whichever branch you want to merge
// git remote remove project-a

node_script.js
const fs = require('fs');

// Define the named pipes
const PYTHON_TO_JS_PIPE = 'python_to_js_pipe';
const JS_TO_PYTHON_PIPE = 'js_to_python_pipe';

// Open pipes for reading and writing
const pythonToJsPipe = fs.openSync(PYTHON_TO_JS_PIPE, 'r');
const jsToPythonPipe = fs.openSync(JS_TO_PYTHON_PIPE, 'w');

// Read a message from Python
const messageFromPython = fs.readFileSync(pythonToJsPipe, 'utf-8').trim();

// Log the received message
console.log('Received message from Python:', messageFromPython);

// Process the message by appending to "Processed:"
const processedMessage = 'Processed: ' + messageFromPython;

// Log the message sent to Python
console.log('Sending message to Python:', processedMessage);

// Send the processed message back to Python
fs.writeSync(jsToPythonPipe, processedMessage);

// Close pipes
fs.closeSync(pythonToJsPipe);
fs.closeSync(jsToPythonPipe);



summarizer.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# This software may be used and distributed according to the terms of the Llama 2 Community License Agreement.

from typing import Optional

import fire

from llama import Llama


import os
import select
import time

def main(
        
    #TODO: queue for intercommunication

    ckpt_dir: str, #TODO: add default dir
    tokenizer_path: str, ##TODO: add default dir
    temperature: float = 0.2,
    top_p: float = 0.95,
    max_seq_len: int = 512,
    max_batch_size: int = 8,
    max_gen_len: Optional[int] = None,
):
    generator = Llama.build(
        ckpt_dir=ckpt_dir,
        tokenizer_path=tokenizer_path,
        max_seq_len=max_seq_len,
        max_batch_size=max_batch_size,
    )

    cache = {}

    instructions = [
        [
            {
                "role": "user",
                "content": "In Bash, how do I list all text files in the current directory (excluding subdirectories) that have been modified in the last month?",
            }
        ],
        [
            {
                "role": "user",
                "content": "What is the difference between inorder and preorder traversal? Give an example in Python.",
            }
        ],
        [
            {
                "role": "system",
                "content": "Provide answers in JavaScript",
            },
            {
                "role": "user",
                "content": "Write a function that computes the set of sums of all contiguous sublists of a given list.",
            }
        ],
    ]

#pipe for intercommunication
IPC_FIFO_NAME_A = "pipe_a"
IPC_FIFO_NAME_B = "pipe_b"

def get_message(fifo):
    '''Read n bytes from pipe. Note: n=24 is an example'''
    return os.read(fifo, 24)

def process_msg(msg):
    '''Process message read from pipe'''
    return msg

    results = generator.chat_completion(
        instructions,  # type: ignore
        max_gen_len=max_gen_len,
        temperature=temperature,
        top_p=top_p,
    )

    for instruction, result in zip(instructions, results):
        for msg in instruction:
            cache[...] = msg["content"] #fn name or the whole hierarchy
            print(f"{msg['role'].capitalize()}: {msg['content']}\n")
        print(
            f"> {result['generation']['role'].capitalize()}: {result['generation']['content']}"
        )
        print("\n==================================\n")


if __name__ == "__main__":
    fire.Fire(main)

    #python pipe execution
    os.mkfifo(IPC_FIFO_NAME_A)  # Create Pipe A
    os.mkfifo(IPC_FIFO_NAME_B)  # Create Pipe B

    try:
        fifo_a = os.open(IPC_FIFO_NAME_A, os.O_RDONLY)  # Pipe A is opened as read-only
        print('Pipe A ready')

        while True:
            try:
                fifo_b = os.open(IPC_FIFO_NAME_B, os.O_WRONLY)
                print("Pipe B ready")
                break
            except:
                # Wait until Pipe B has been initialized
                time.sleep(1)

        try:
            poll = select.poll()
            poll.register(fifo_a, select.POLLIN)

            try:
                while True:
                    if (fifo_a, select.POLLIN) in poll.poll(1000):  # Poll every 1 sec
                        msg = get_message(fifo_a)                   # Read from Pipe A
                        msg = process_msg(msg)                      # Process Message
                        os.write(fifo_b, msg)                       # Write to Pipe B

                        print('----- Received from JS -----')
                        print("    " + msg.decode("utf-8"))
            finally:
                poll.unregister(fifo_a)
        finally:
            os.close(fifo_a)
    finally:
        os.remove(IPC_FIFO_NAME_A)
        os.remove(IPC_FIFO_NAME_B)
